### ISSUE TITLE
Feature: Dark mode

### ISSUE THREAD

**Issue by ShamblerBishop** [WinMerge/winmerge#396]
Working with WinMerge on a 32" 4K monitor, alt-tabbing between dark-mode Visual Studio, and a full screen WinMerge - is pretty jarring/searing on my eyes - be nice to have a dark mode, like VS!

**Comment by ShamblerBishop** [comment]
Ah,  duplicate.

**Comment by ikappas** [comment]
@ShamblerBishop I don't think this should be a duplicate of Dark theme since Dark mode is *NOT* Dark theme.

As implemented in latest Windows 10, Dark mode refers to adjusting the whole application look an feel to be presented in dark colors including main window, menus, icons instead of just theming the content area.

@sdottaka Perhaps this should be re-opened/re-considered?

**Comment by Neustradamus** [comment]
Any news about it?

**Comment by sdottaka** [comment]
So far I haven't done anything about dark mode.

I expected to improve Dark Mode support for Win32 applications on Windows 11, but I haven't seen any improvement.

Currently, I'm not very enthusiastic because dark mode can only be achieved using an undocumented API. Pull Requests are welcome.

**Comment by PwrSrg** [comment]
+1  This is desperately needed.  The "dark" theme (which is hideous btw) is **_not_** the same thing as **dark MODE**.

But for the PRICE, I cannot complain! 😊  Though I _am_ about to switch to **Beyond Compare** which is rolling out Dark Mode in the upcoming v5 release. 🌑

**Comment by andresbnd** [comment]
+1 Please we need that! It will be amazing

**Comment by rigiddesign** [comment]
Considering WinMerge is free software, perhaps it would be worth the author offering up a crowdfunding option to try and fund the development of this.

**Comment by Laurens-Bamtang** [comment]
Hi, I guess the undocumented API calls are these? 
https://github.com/ysc3839/win32-darkmode/tree/master
I tested this on my PC and it didn't work, and my Windows 10 is in theory "updated"
Since the program uses MFC that doesn't support dark mode at all I guess supporting it is out of the question, isn't it?

**Comment by flappysquirrel** [comment]
I just went down this rabbithole.

Free alternatives with an actual dark UI:

Meld (meldmerge.org)

P4Merge (perforce.com)

Actually, the perforce website gives off strong Oracle vibes, so I'm not too sure about the 'free' part.

**Comment by oeloo** [comment]
I like WinMerge, but the lack of dark mode makes it unusable for me: when WIndows OS is in dark mode, using WinMerge makes white flashes in your eyes each time you open its window, which is killing the eyes.
WinMerge definitely needs a dark mode. Almost all applications, now end 2023, have a light/dark mode automatically switching with Windows OS light/dark mode.

**Comment by Tamarindo94** [comment]
This is a much needed feature!

**Comment by kursad-k** [comment]
> I just went down this rabbithole.
> > Free alternatives with an actual dark UI:
> 
> Meld (meldmerge.org)
> 


Meld does not look good on win 11 with Dark Mode. 

![image](https://github.com/WinMerge/winmerge/assets/67970138/d637693d-0f39-4638-9e01-e5c715d627ea)

**Comment by luckyluc74** [comment]
for some code inspiration, notepad++ implemented [dark mode](https://github.com/search?q=repo%3Anotepad-plus-plus%2Fnotepad-plus-plus%20dark&type=code), maybe this helps

**Comment by bamtang-dev** [comment]
> for some code inspiration, notepad++ implemented [dark mode](https://github.com/search?q=repo%3Anotepad-plus-plus%2Fnotepad-plus-plus%20dark&type=code), maybe this helps

interesting, but do they use MFC? I quickly checked their source code and I can't find if they are using MFC that is the main blocker.

**Comment by Pysis868** [comment]
Microsoft Foundation Class (MFC) Library

https://learn.microsoft.com/en-us/cpp/mfc/reference/creating-a-forms-based-mfc-application?view=msvc-170
https://learn.microsoft.com/en-us/cpp/mfc/reference/creating-an-mfc-application?view=msvc-170
https://learn.microsoft.com/en-us/cpp/mfc/mfc-desktop-applications?view=msvc-170

https://learn.microsoft.com/en-us/cpp/mfc/mfc-library-versions?view=msvc-170

Static library files named `(N)(U)AFX#(D).LIB`.
Shared library files named `MFC#(U)(D).DLL`.

Found an example:

https://learn.microsoft.com/en-us/cpp/windows/redistributing-the-mfc-library?view=msvc-170

`mfc140.dll`

.../WinMerge/WinMergeU.exe
https://www.virustotal.com/gui/file/975a092a2249ecb74e641fde94e51b01c304886a44c9114c35fd956131c47592/summary

Based on the VT report, it is probably a Win32 App, before MFC?  Based on those several `*32.dll` files implementing parts of the complete API.

[Reduce Win32API type usage in code](https://github.com/WinMerge/winmerge/commit/f2feefc563f374e290de55fbbcea94416d292d8d)

Unrelated:
https://malwaretips.com/blogs/vcomp140-dll-what-it-is-how-to-fix-errors/

https://github.com/WinMerge/winmerge/blob/master/Src/Merge.vcxproj
`Link` sections, `DelayLoadDLLs` entries have `comdlg32.dll`.  That may be the main one.

**Comment by AutumnGamer** [comment]
+1 I also wanted to ask for dark mode that includes the whole UI! I would much prefer a dark mode that goes with the default darkmode native to windows 10, which is mostly black, instead of the colors currently offered in the limited dark themes. For example, Notepad3 added that as an option called 'Windows Dark Mode' and it works really well and is much calmer to the eyes!

**Comment by MkazemAkhgary** [comment]
I hope the developers can prioritize this issue. Overall, I think this app is great, and having the proper dark theme with black borders would make it even better. right now dark theme has very sharp white borders which makes it pretty much unusable

**Comment by luckyluc74** [comment]
Maybe this is also helpful [Win32-darkmode example  ](https://github.com/ysc3839/win32-darkmode)

**Comment by gabriel-vanca** [comment]
Any news on this?

**Comment by Jack251970** [comment]
Any news?

**Comment by stanio** [comment]
> Any news?

Have you tried the latest release ([2.16.44](https://github.com/WinMerge/winmerge/releases/tag/v2.16.44)) that includes:

>   -   Add Support for Custom System Color Overrides (PR #2376)

?

<img alt="Options" src="https://github.com/user-attachments/assets/6bd4c331-d547-4760-8484-882d6b5595b0" width="50%">

![WinMerge Dark](https://github.com/user-attachments/assets/bce6edd9-9179-4bc7-8e8a-0cb35d441e6f)

**Comment by Jack251970** [comment]
> > Any news?
> 
> Have you tried the latest release ([2.16.44](https://github.com/WinMerge/winmerge/releases/tag/v2.16.44)) that includes:
> 
> > * Add Support for Custom System Color Overrides (PR [Add Support for Custom System Color Overrides #2376](https://github.com/WinMerge/winmerge/pull/2376))
> 
> ?
> 
> <img alt="Options" width="50%" src="https://private-user-images.githubusercontent.com/1247730/393735025-6bd4c331-d547-4760-8484-882d6b5595b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3Mjg0OTIsIm5iZiI6MTczMzcyODE5MiwicGF0aCI6Ii8xMjQ3NzMwLzM5MzczNTAyNS02YmQ0YzMzMS1kNTQ3LTQ3NjAtODQ4NC04ODJkNmI1NTk1YjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDlUMDcwOTUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTJmYjkxMGFhMGZiNzZlYzNkMTkwZDM5NWFlOTg1MTA3NTczMzU1OGFjNWRmOGQzZjk4MjQ3NTkzM2Q3ZGEyMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.hVjI6VUFkru08tjVtOt6OZKsjbBUoJWEWWDC2p_8Pi4">
> ![WinMerge Dark](https://private-user-images.githubusercontent.com/1247730/393735370-bce6edd9-9179-4bc7-8e8a-0cb35d441e6f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3Mjg0OTIsIm5iZiI6MTczMzcyODE5MiwicGF0aCI6Ii8xMjQ3NzMwLzM5MzczNTM3MC1iY2U2ZWRkOS05MTc5LTRiYzctOGU4YS0wY2IzNWQ0NDFlNmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDlUMDcwOTUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjJiNDlmMDNhNjI4Y2MyYTZhZTkzYzA3M2NhNjE2MzllODZkOTlhYTJlMDZkNjI5YTAxNzA3ZjczZjU0OWIyOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.DHfQFkRXAsQ4IH1_gMUA-DguoAjFLg6x2cWAA3NhTUA)

Thank you! I find it! Just forget to update to the latest version.

**Comment by rkodey** [comment]
FWIW:
A quick search found this additional dark mode theme:  https://github.com/liqube/liqube-dark-winmerge
And then I created one of my own:  https://github.com/rkodey/winmerge-visual-studio-dark

WinMerge 2.16.44 added the ability to override some of the system colors.  Big improvement!

**Comment by rigiddesign** [comment]
> FWIW: A quick search found this additional dark mode theme: https://github.com/liqube/liqube-dark-winmerge And then I created one of my own: https://github.com/rkodey/winmerge-visual-studio-dark
> 
> Big improvement with WinMerge 2.16.44! Thanks!!

Nice one; thank you.

Is it possible for the blank UI, i.e, before you select any files, to be darker? Anyone know how to do this?

![image](https://github.com/user-attachments/assets/437e3329-fee0-42e6-a2ab-f2f809b63051)

**Comment by luckyluc74** [comment]
@rkodey Nice job for the new dark theme and also all the WinMerge developers for making this possible with WinMerge 2.16.44!  

Almost there :)
Hopefully we can also get menu/context menu dark and scroll bars dark in dark mode 

This is example code for dark mode scrollbars in [Win32](https://gist.github.com/rounk-ctrl/b04e5622e30e0d62956870d5c22b7017#dark-scrollbars)
 
This is example for dark mode context menus in [win32](https://gist.github.com/rounk-ctrl/b04e5622e30e0d62956870d5c22b7017#setpreferredappmodein-preferredappmode-appmode)

**Comment by luckyluc74** [comment]
With the following patch on CMainFrame (cpp/h) I was able to get Dark windows menu and context menu.
Only thing is that it does not change dark/non dark mode when changing settings in options, only when WinMerge is restarted.

![Image](https://github.com/user-attachments/assets/3482954d-3f96-45df-9a62-9b0bf16ebe4c)

![Image](https://github.com/user-attachments/assets/701ce372-0458-44f8-a9eb-6449dcab449c)

This is the patch
```
 Src/MainFrm.cpp | 21 +++++++++++++++++++++
 Src/MainFrm.h   |  9 +++++++++
 2 files changed, 30 insertions(+)

diff --git a/Src/MainFrm.cpp b/Src/MainFrm.cpp
index 15545a374..c683d0b5b 100644
--- a/Src/MainFrm.cpp
+++ b/Src/MainFrm.cpp
@@ -411,8 +411,29 @@ BOOL CMainFrame::PreCreateWindow(CREATESTRUCT& cs)
 	return bRes;
 }
 
+void CMainFrame::SetPreferredAppMode(PreferredAppMode preferredAppMode)
+{
+	HMODULE hUxtheme = LoadLibraryW(L"uxtheme.dll");
+	if (!hUxtheme) return;
+
+	typedef PreferredAppMode(WINAPI* fnSetPreferredAppMode)(PreferredAppMode appMode); // ordinal 135, available since Windows 10 build version 18334
+	auto SetPreferredAppMode = reinterpret_cast<fnSetPreferredAppMode>(
+		GetProcAddress(hUxtheme, MAKEINTRESOURCEA(135))
+		);
+
+	if (SetPreferredAppMode)
+	{
+		SetPreferredAppMode(preferredAppMode);
+	}
+
+	FreeLibrary(hUxtheme);
+}
+
 int CMainFrame::OnCreate(LPCREATESTRUCT lpCreateStruct)
 {
+	if (IsWin10_OrGreater())
+		SetPreferredAppMode(GetOptionsMgr()->GetBool(OPT_SYSCOLOR_HOOK_ENABLED) ? ForceDark : ForceLight);
+
 	if (__super::OnCreate(lpCreateStruct) == -1)
 		return -1;
 
diff --git a/Src/MainFrm.h b/Src/MainFrm.h
index d2b5656a2..ea8226da2 100644
--- a/Src/MainFrm.h
+++ b/Src/MainFrm.h
@@ -43,6 +43,14 @@ class DirWatcher;
 
 typedef std::shared_ptr<TempFile> TempFilePtr;
 
+typedef enum _PreferredAppMode {
+	Default,
+	AllowDark,
+	ForceDark,
+	ForceLight,
+	Max
+} PreferredAppMode;
+
 // typed lists (homogenous pointer lists)
 typedef CTypedPtrList<CPtrList, COpenDoc *> OpenDocList;
 typedef CTypedPtrList<CPtrList, CMergeDoc *> MergeDocList;
@@ -443,6 +451,7 @@ protected:
 	DECLARE_MESSAGE_MAP()
 
 private:
+	void SetPreferredAppMode(PreferredAppMode preferredAppMode);
 	void addToMru(const tchar_t* szItem, const tchar_t* szRegSubKey, UINT nMaxItems = 20);
 	OpenDocList &GetAllOpenDocs();
 	MergeDocList &GetAllMergeDocs();

```

**Comment by luckyluc74** [comment]
Small update

On my local branch of master of Winmerge, menu and context menus look correct in dark mode.
Fix for lightgrey border was to change following code in Merge.cpp , to IsWin10 instead of IsWIn11. FYI I am using windows 10.

![Image](https://github.com/user-attachments/assets/9d6fc95e-547c-4bee-b2f6-670838fa6cab)


I temporary removed ( | WS_HSCROLL | WS_VSCROLL ) in MergeEditSplitterView.cpp and then the scroll bars are also dark, but the scollbars then do not work except for mousewheel. How to make the scrollbars working in Dark mode is a big question mark right now for me. Also dialogs are all light, but that is another challenge. 


And then it looks like this

![Image](https://github.com/user-attachments/assets/543ce360-2142-4fca-afc4-e58af26be6c7)

I used to program in win32 MFC c/c++ but that was in the late 90s :)

**Comment by luckyluc74** [comment]
> > FWIW: A quick search found this additional dark mode theme: https://github.com/liqube/liqube-dark-winmerge And then I created one of my own: https://github.com/rkodey/winmerge-visual-studio-dark
> > Big improvement with WinMerge 2.16.44! Thanks!!
> 
> Nice one; thank you.
> 
> Is it possible for the blank UI, i.e, before you select any files, to be darker? Anyone know how to do this?
> 
> ![image](https://github.com/user-attachments/assets/437e3329-fee0-42e6-a2ab-f2f809b63051)


In file 'MainFrm.h'  in class CMDIClient you can add the following

```
			case WM_ERASEBKGND:
				{
					CRect rect;
					GetMainFrame()->GetClientRect(&rect);
					CBrush brush(GetSysColor(COLOR_BACKGROUND)); 
					GetMainFrame()->GetDC()->FillRect(&rect, &brush);
					return TRUE;
				}
				break;
```

![Image](https://github.com/user-attachments/assets/bdf33881-45b0-4344-aed8-42d3052af78e)

and then background looks like this

![Image](https://github.com/user-attachments/assets/711c9244-5e6e-4834-9abc-e926525c7a6c)

**Comment by rkodey** [comment]
Looks like great progress.  Are you planning a PR?

**Comment by lucatmedmij** [comment]
> Looks like great progress. Are you planning a PR?

Thanks! Yes, I am planning a PR. I already have the dark scrollbars working as should be. Need to figure out how to switch to dark and non dark mode at startup, depending on theme.

**Comment by luckyluc74** [comment]
> Looks like great progress. Are you planning a PR?

Thanks! Yes, I am planning to create a PR. I already have the dark scrollbars working as should be. Need to figure out how to switch to dark and non dark mode at startup, depending on theme. 

![Image](https://github.com/user-attachments/assets/b9ff5fb2-0eba-4064-9d7b-666dc3e62cf0)

**Comment by luckyluc74** [comment]
Update, first attempt to make dialog box "Patch Generator" dark.

Combobox, button, checkbox and dialog itself are dark. Only groupbox and dialog title are still not dark.

![Image](https://github.com/user-attachments/assets/802f6e27-86b6-42d0-bb2b-0e5d3b8b73bf)

**Comment by rigiddesign** [comment]
Amazing! Thank you, @luckyluc74.

Let us know when this is available, please 💜

**Comment by luckyluc74** [comment]
It is a lot of work to change all dialog boxes to dark mode, so this will take some time.

Preview of Dark mode about box. For now I use filter reverse color and then 180 degree hue to make dark mode image.

![Image](https://github.com/user-attachments/assets/15cbd171-bcd9-4abc-851b-03ea1bb156bb)

**Comment by luckyluc74** [comment]
This is Open dialog (WIP).  More challenging one todo.

![Image](https://github.com/user-attachments/assets/12145465-fafb-425b-9336-b0057372966c)

**Comment by rigiddesign** [comment]
> This is Open dialog (WIP). More challenging one todo.
> 
> ![Image](https://github.com/user-attachments/assets/2c1ac590-6657-47e4-8a04-9c571bad292e)

It is better than being blinded by the light, though!

**Comment by ozone10** [comment]
I think I don't like MFC, anyway here is proof of concept.
https://github.com/ozone10/winmerge/actions/runs/14621047024

It can be unstable since I am not familiar with winmerge code and its submodules structures.
Basic dialogs should work, if there are some dialogs in light mode it is probably from external (submodule) code.

I won't be working on this soon for now, so don't expect PR. I may do rebase on latest master if I have time.

![Image](https://github.com/user-attachments/assets/1099699e-e29b-4157-a7a8-f21f7a66b371)

I would like to thank @ luckyluc74 for code for scrollbars and CMDIClient background, and since I am not used to MFC it really helped.

**Comment by ozone10** [comment]
I had little time to do some modifications. I have tested only x64 release variant. WinMerge building configuration is very complex, and unfortunately I have no time to check other configurations (x86, ARM, ARM64 and debug, test).
Below is "DarkModeLib" variant branch which move dark mode code to external static library to be able to use it in other WinMerge  external dependencies.
https://github.com/ozone10/winmerge/actions?query=branch%3ADarkModeLib

Currently it is also used  for crystal edit dialogs (find, replace, and marker) since git submodules is not used for it.

![Image](https://github.com/user-attachments/assets/2fbfe472-574e-4c3f-a98a-856cbf11d1a4)

![Image](https://github.com/user-attachments/assets/5734c208-5e3a-45bf-97bb-3502f2037e22)

![Image](https://github.com/user-attachments/assets/e423a98a-7043-4678-a5a4-9fc9585ff1be)

![Image](https://github.com/user-attachments/assets/1c38d47d-40ac-4431-b895-7e495f818985)

What is not currently in dark mode: webpage and image diff; 3 Frhed binary setting dialogs. I won't be working on these.

Automatic change of dark mode on system is not implemented, WinMerge will remain same until restart.

Also since this is based on code from https://github.com/ozone10/7zip-Dark7zip/tree/main/DarkMode
You can configure WinMerge dark mode colors with values from https://github.com/ozone10/7zip-Dark7zip?tab=readme-ov-file#config with `winMergeDark.ini` file

**Comment by ozone10** [comment]
Just in case, if someone want PR, before doing it, I need some feedback and testing. 

If you can build then mainly compilations for other than vs2022 (sln files are not updated for both vs2017 and vs2019 yet) and other than x64 release variant for branch https://github.com/ozone10/winmerge/tree/DarkModeLib. Take a note that dark mode is currently enabled only for x64 and arm64 builds, and will work fully only on supported Windows 10 and 11 builds. On older OS from Vista only partially, no dark scrollbars, tooltips, listview header, and popup menus, ...

For users check if dark mode is applied correctly, except for webpage and image diff; 3 Frhed binary setting dialogs.
I am using only basic sets of winmerge functions so I have no idea if there is some "special" dialog, which is used only by advanced users. 

You can check both branch but preferably "DarkModeLib" one.
https://github.com/ozone10/winmerge/tree/DarkMode
https://github.com/ozone10/winmerge/tree/DarkModeLib

I will be also busy so it won't be soon.

**Comment by dcog989** [comment]
@ozone10 - I saw your comment on #2665. I'll try and build DarkModeLib tomorrow and feedback. [comment as reminder]

**Comment by dcog989** [comment]
@ozone10 - I have a PR waiting and I'm not able to fork your branch to my repo (or at least I don't know how to). I'll come back to this once PR is merged - or rejected!

**Comment by ozone10** [comment]
@dcog989 
you don't fork forked repository, instead you add remote
1. open git bash in your local winmerge repository 
2. `git remote add ozone10 https://github.com/ozone10/winmerge.git` 
3. `git fetch ozone10`
4. `git switch DarkModeLib`
5. now you should have DarkModeLib branch on your local repository

**Comment by dcog989** [comment]
@ozone10 - thanks. My 'git fu' is weak! I'll try that tomorrow.

**Comment by dcog989** [comment]
@ozone10 - apologies for delay, life got in the way.

Your git CLI worked fine and I've built your branch. Feedback:


1. white checkboxes, plus this blue / green tint in Options - ![Image](https://github.com/user-attachments/assets/5bff4896-aecc-4392-be07-a9e81cdbb06a)
2. more white checkboxes - ![Image](https://github.com/user-attachments/assets/089d9a1d-6adc-47c0-849f-f98342583dcf)
![Image](https://github.com/user-attachments/assets/824644b7-9ed5-4511-bc07-0b91dcff9b1f)
3. Open files dialog - light grey background would be better dark - ![Image](https://github.com/user-attachments/assets/aabe960f-9731-48f7-a7fe-6881ae885ba7)
4. Tools... Generate Patch - light grey bg again - ![Image](https://github.com/user-attachments/assets/090056fa-fc22-4405-bc90-4ba096df59a5)
5. black bullet point on dark background - ![Image](https://github.com/user-attachments/assets/b0d78f58-9656-477f-8fed-f4f9dbfab5e2)

Plus, these dialogs:

- File... Print 
- File... Page setup
- File... Print Preview
- View... Syntax Highlight
- Help... About

---

But overall, excellent work. A huge improvement on what we have currently. 👍

**Comment by ozone10** [comment]
@dcog989 

thank you!

> 1\. white checkboxes, plus this blue / green tint in Options -

> 2\. more white checkboxes -

For checkboxes this is current limitation of Windows 11 theme, and owner draw is something I don't want to do now.
For blue / green tint could you be more specific.

> 3\. Open files dialog - light grey background would be better dark -

> 4\. Tools... Generate Patch - light grey bg again -

> * Help... About

This controlled by color scheme, I will try to override it. 

> 5\. black bullet point on dark background -

I will check it.

> * File... Print
> 
> * File... Page setup
> 
> * File... Print Preview
> 
> * View... Syntax Highlight

I probably won’t do them, but they may need to use an MFC-derived class and I am not proficient with it.

**Comment by dcog989** [comment]
Excellent. Please let me know if I can help further.

**Comment by ozone10** [comment]
Fixed or improved:
-  File... Page setup -  fully in dark mode
- View... Syntax Highlight - two columns mode will unfortunately disable dark mode visuals, unless we use owner draw menu, which has thick light border, one column mode may clash with menu bar under it.
- black bullet point on dark background - bullet point with other than black color will not be dot ???
- Font dialog - for some reason it will not process `WM_CTLCOLORLISTBOX` for combobox listboxes

Need MFC derived class - which I won't do
- Choose color dialog
- Print dialog
- Print preview

Consideration
- replace color button client edge with simple flat border - currently only button in system has it

TODO
- sync some colors from from system option with dark mode defined ones

**Comment by dcog989** [comment]
Confirmed, looks good.

**Comment by oeloo** [comment]
Is there a command line to turn WinMerge to dark mode / light mode?

**Comment by ozone10** [comment]
> Is there a command line to turn WinMerge to dark mode / light mode?

For color scheme I don't know.
For my dark mode implementation, no. It does not have GUI settings or command lines, which I will leave to other to implement.

**Comment by oeloo** [comment]
As there is no Windows System Auto dark/light mode implemented in WinMerge, we need at least a command line to switch between dark and light mode, to manage this missing feature.

**Comment by dcog989** [comment]
No, you can switch to dark mode by selecting a dark theme or customising colours in Settings. Some chrome is already done, ozone10 is converting most of the remainder

**Comment by MkazemAkhgary** [comment]
when will this get released?

**Comment by tamr** [comment]
Any way to make a lighter tone diff colors in Location Pane?

**Comment by ozone10** [comment]
@sdottaka 
good news I was able to get permission to change license to MPL-2.0 which is compatible with GPLv2 (https://github.com/ozone10/darkmodelib/issues/1).
I will try to prepare initial PR with basic support soon.

**Comment by sdottaka** [comment]
> [@sdottaka](https://github.com/sdottaka) good news I was able to get permission to change license to MPL-2.0 which is compatible with GPLv2 ([ozone10/darkmodelib#1](https://github.com/ozone10/darkmodelib/issues/1)). I will try to prepare initial PR with basic support soon.

Great news! Looking forward to your PR with the initial support.

**Comment by ozone10** [comment]
WIP
https://github.com/ozone10/winmerge/tree/DarkModeInitialSupport
https://github.com/ozone10/winmerge/actions/runs/16057792447

Should work for vs2022. For older vs2017 and vs2019 maybe only x86 and arm. Also I need to figure how are git submodules handled here, since it does not seems it is used in github actions via checkout action.
